### ISSUE TITLE
LDAP not authenticating to openldap

### ISSUE THREAD

**Issue by dnhodgson** [opensearch-project/security#985]
Currently trying to setup LDAP authentication against OD 1.11.0 and am unable to get it work.

I increased logging to try to troubleshoot but still can't figure out why this isn't working.

Trace logs from attempting authentication with a binddn user set:
```
[2021-01-16T16:23:01,924][TRACE][c.a.d.a.l.b.LDAPAuthorizationBackend] [es-dev-opendistro-es-client-6c696b5744-ljn5w] Connect to ldap.cluster.local:389
[2021-01-16T16:23:01,924][TRACE][c.a.d.a.l.b.LDAPAuthorizationBackend] [es-dev-opendistro-es-client-6c696b5744-ljn5w] Connect to ldap://ldap.cluster.local:389
[2021-01-16T16:23:01,924][DEBUG][c.a.d.a.l.b.LDAPAuthorizationBackend] [es-dev-opendistro-es-client-6c696b5744-ljn5w] verifyHostname true:
[2021-01-16T16:23:01,925][DEBUG][c.a.d.a.l.b.LDAPAuthorizationBackend] [es-dev-opendistro-es-client-6c696b5744-ljn5w] trustall false:
[2021-01-16T16:23:01,930][DEBUG][c.a.d.a.l.b.LDAPAuthorizationBackend] [es-dev-opendistro-es-client-6c696b5744-ljn5w] Use Trust-/Keystore to secure communication with LDAP server (client auth is false)
[2021-01-16T16:23:01,930][DEBUG][c.a.d.a.l.b.LDAPAuthorizationBackend] [es-dev-opendistro-es-client-6c696b5744-ljn5w] trustStoreAliases: null, keyStoreAlias: null
[2021-01-16T16:23:01,930][DEBUG][c.a.d.a.l.b.LDAPAuthorizationBackend] [es-dev-opendistro-es-client-6c696b5744-ljn5w] enabled ssl/tls protocols for ldaps [TLSv1.2, TLSv1.1]
[2021-01-16T16:23:01,930][DEBUG][c.a.d.a.l.b.LDAPAuthorizationBackend] [es-dev-opendistro-es-client-6c696b5744-ljn5w] Connect timeout: PT5S/ResponseTimeout: PT0S
[2021-01-16T16:23:01,931][DEBUG][c.a.d.a.l.b.LDAPAuthorizationBackend] [es-dev-opendistro-es-client-6c696b5744-ljn5w] bindDn uid=binduser,ou=People,dc=cluster,dc=local, password ****
[2021-01-16T16:23:01,931][DEBUG][c.a.d.a.l.b.LDAPAuthorizationBackend] [es-dev-opendistro-es-client-6c696b5744-ljn5w] Will perform simple bind with bind dn
[2021-01-16T16:23:01,937][WARN ][c.a.d.a.l.b.LDAPAuthorizationBackend] [es-dev-opendistro-es-client-6c696b5744-ljn5w] Unable to connect to ldapserver ldap.cluster.local:389 due to [org.ldaptive.provider.ConnectionException@1379678003::resultCode=INVALID_CREDENTIALS, matchedDn=null, responseControls=null, referralURLs=null, messageId=-1, message=javax.naming.AuthenticationException: [LDAP: error code 49 - Invalid Credentials], providerException=javax.naming.AuthenticationException: [LDAP: error code 49 - Invalid Credentials]]. Try next.
[2021-01-16T16:23:01,938][DEBUG][c.a.d.a.l.b.LDAPAuthorizationBackend] [es-dev-opendistro-es-client-6c696b5744-ljn5w] Unable to connect to ldapserver due to 
org.ldaptive.provider.ConnectionException: javax.naming.AuthenticationException: [LDAP: error code 49 - Invalid Credentials]
```

Trace logs from an anonymous bind
```
[2021-01-16T17:21:43,953][TRACE][c.a.d.a.l.b.LDAPAuthorizationBackend] [es-dev-opendistro-es-client-6c696b5744-s7h2t] Connect to ldap.cluster.local:389
[2021-01-16T17:21:43,954][TRACE][c.a.d.a.l.b.LDAPAuthorizationBackend] [es-dev-opendistro-es-client-6c696b5744-s7h2t] Connect to ldap://ldap.cluster.local:389
[2021-01-16T17:21:43,954][DEBUG][c.a.d.a.l.b.LDAPAuthorizationBackend] [es-dev-opendistro-es-client-6c696b5744-s7h2t] verifyHostname true:
[2021-01-16T17:21:43,954][DEBUG][c.a.d.a.l.b.LDAPAuthorizationBackend] [es-dev-opendistro-es-client-6c696b5744-s7h2t] trustall false:
[2021-01-16T17:21:43,957][DEBUG][c.a.d.a.l.b.LDAPAuthorizationBackend] [es-dev-opendistro-es-client-6c696b5744-s7h2t] Use Trust-/Keystore to secure communication with LDAP server (client auth is false)
[2021-01-16T17:21:43,957][DEBUG][c.a.d.a.l.b.LDAPAuthorizationBackend] [es-dev-opendistro-es-client-6c696b5744-s7h2t] trustStoreAliases: null, keyStoreAlias: null
[2021-01-16T17:21:43,957][DEBUG][c.a.d.a.l.b.LDAPAuthorizationBackend] [es-dev-opendistro-es-client-6c696b5744-s7h2t] enabled ssl/tls protocols for ldaps [TLSv1.2, TLSv1.1]
[2021-01-16T17:21:43,957][DEBUG][c.a.d.a.l.b.LDAPAuthorizationBackend] [es-dev-opendistro-es-client-6c696b5744-s7h2t] Connect timeout: PT5S/ResponseTimeout: PT0S
[2021-01-16T17:21:43,957][DEBUG][c.a.d.a.l.b.LDAPAuthorizationBackend] [es-dev-opendistro-es-client-6c696b5744-s7h2t] bindDn null, password <not set>
[2021-01-16T17:21:43,957][DEBUG][c.a.d.a.l.b.LDAPAuthorizationBackend] [es-dev-opendistro-es-client-6c696b5744-s7h2t] Will perform anonymous bind because no bind dn is given
[2021-01-16T17:21:44,035][DEBUG][c.a.d.a.l.b.LDAPAuthorizationBackend] [es-dev-opendistro-es-client-6c696b5744-s7h2t] Opened a connection, total count is now 1
[2021-01-16T17:21:44,051][DEBUG][c.a.d.a.l.b.LDAPAuthenticationBackend] [es-dev-opendistro-es-client-6c696b5744-s7h2t] Results for LDAP search for user1 in base _legacyConfig:
[]
[2021-01-16T17:21:44,051][TRACE][c.a.d.a.l.b.LDAPAuthenticationBackend] [es-dev-opendistro-es-client-6c696b5744-s7h2t] Try to authenticate dn uid=user1,ou=People,dc=cluster,dc=local
[2021-01-16T17:21:44,051][DEBUG][c.a.d.a.l.b.LDAPAuthorizationBackend] [es-dev-opendistro-es-client-6c696b5744-s7h2t] Closed a connection, total count is now 0
[2021-01-16T17:21:44,052][DEBUG][c.a.d.a.l.b.LDAPAuthorizationBackend] [es-dev-opendistro-es-client-6c696b5744-s7h2t] bindDn uid=user1,ou=People,dc=cluster,dc=local, password ****
[2021-01-16T17:21:44,059][DEBUG][c.a.d.a.l.b.LDAPAuthenticationBackend] [es-dev-opendistro-es-client-6c696b5744-s7h2t] Unable to authenticate user due to 
org.ldaptive.provider.ConnectionException: javax.naming.AuthenticationException: [LDAP: error code 49 - Invalid Credentials]
```

Testing to make sure both my user accounts/passwords work from the client container, as well as anonymous binding.

```
[root@es-dev-opendistro-es-client-6c696b5744-ljn5w elasticsearch]# ldapsearch -ZZ -H ldap://ldap.cluster.local -x -b "ou=People,dc=cluster,dc=local" '(&(objectClass=posixAccount)(uid=user1))' -D uid=binduser,ou=People,dc=cluster,dc=local -W
Enter LDAP Password: 
# extended LDIF

[root@es-dev-opendistro-es-client-6c696b5744-cb682 elasticsearch]# ldapsearch -ZZ -H ldap://ldap.cluster.local -x -b "ou=People,dc=cluster,dc=local" '(&(objectClass=posixAccount)(uid=user1))' -D uid=user1,ou=People,dc=cluster,dc=local -W
Enter LDAP Password: 
# extended LDIF

[root@es-dev-opendistro-es-client-6c696b5744-cb682 elasticsearch]# ldapsearch -ZZ -H ldap://ldap.cluster.local -x -b "ou=People,dc=cluster,dc=local" '(&(objectClass=posixAccount)(uid=user1))'
# extended LDIF
```

Config File
```
_meta:
  type: "roles"
  config_version: 2
config:
  dynamic:
    http:
      anonymous_auth_enabled: false
    authc:
      basic_internal_auth_domain:
        description: "Authenticate via HTTP Basic against internal users database"
        http_enabled: true
        transport_enabled: false
        order: 4
        http_authenticator:
          type: basic
          challenge: true
        authentication_backend:
          type: intern
      ldap:
        http_enabled: true
        transport_enabled: false
        order: 1
        http_authenticator:
          type: basic
          challenge: true
        authentication_backend:
          type: ldap
          config:
            enable_ssl: false
            enable_start_tls: true
            enable_ssl_client_auth: false
            verify_hostnames: true
            hosts:
              - ldap.cluster.local:389
            userbase: ou=People,dc=cluster,dc=local
            usersearch: '(uid={0})' 
            username_attribute: uid
    authz:
      roles_from_myldap:
        http_enabled: true
        transport_enabled: false
        authorization_backend:
          type: ldap
          config:
            enable_ssl: false
            enable_start_tls: true
            enable_ssl_client_auth: false
            verify_hostnames: true
            hosts:
              - ldap.cluster.local:389
            userbase: ou=People,dc=cluster,dc=local
            usersearch: '(uid={0})'
            username_attribute: uid
            rolebase: ou=Group,dc=cluster,dc=local
            rolesearch: '(memberUid={0})'
            rolename: cn
            skip_users:
              - kibanaserver
              - admin
```

I have also tried using enable_ssl instead of starttls and port 636 but that produces the same results.

**Comment by smortex** [comment]
Similar setup and similar issue.

On the ldap server side, STARTTLS is missing before authentication:

```
août 10 18:23:20 vps-b12aeabe slapd[17127]: conn=1027 fd=12 ACCEPT from IP=202.90.71.171:53850 (IP=0.0.0.0:389)
août 10 18:23:20 vps-b12aeabe slapd[17127]: conn=1027 op=0 BIND dn="cn=opensearch,ou=services,dc=example,dc=com" method=128
août 10 18:23:20 vps-b12aeabe slapd[17127]: conn=1027 op=0 RESULT tag=97 err=49 text=
août 10 18:23:20 vps-b12aeabe slapd[17127]: conn=1027 fd=12 closed (connection lost)
```

When connecting with `ldapsearch -ZZ -x -D cn=opensearch,ou=services,dc=example,dc=com -w secret  -b ou=people,dc=example,dc=com '(uid=romain)'`, I can see a STARTLS before a successful authentication:

```
août 10 18:23:09 vps-b12aeabe slapd[17127]: conn=1026 fd=12 ACCEPT from IP=202.90.71.171:64584 (IP=0.0.0.0:389)
août 10 18:23:09 vps-b12aeabe slapd[17127]: conn=1026 op=0 EXT oid=1.3.6.1.4.1.1466.20037
août 10 18:23:09 vps-b12aeabe slapd[17127]: conn=1026 op=0 STARTTLS
août 10 18:23:09 vps-b12aeabe slapd[17127]: conn=1026 op=0 RESULT oid= err=0 text=
août 10 18:23:09 vps-b12aeabe slapd[17127]: conn=1026 fd=12 TLS established tls_ssf=256 ssf=256
août 10 18:23:09 vps-b12aeabe slapd[17127]: conn=1026 op=1 BIND dn="cn=opensearch,ou=services,dc=example,dc=com" method=128
août 10 18:23:09 vps-b12aeabe slapd[17127]: conn=1026 op=1 BIND dn="cn=opensearch,ou=services,dc=example,dc=com" mech=SIMPLE ssf=0
août 10 18:23:09 vps-b12aeabe slapd[17127]: conn=1026 op=1 RESULT tag=97 err=0 text=
août 10 18:23:10 vps-b12aeabe slapd[17127]: conn=1026 op=2 SRCH base="ou=people,dc=example,dc=com" scope=2 deref=0 filter="(uid=romain)"
août 10 18:23:10 vps-b12aeabe slapd[17127]: conn=1026 op=2 SEARCH RESULT tag=101 err=0 nentries=1 text=
août 10 18:23:10 vps-b12aeabe slapd[17127]: conn=1026 op=3 UNBIND
août 10 18:23:10 vps-b12aeabe slapd[17127]: conn=1026 fd=12 closed
```

Authentication fail because the client has not started a TLS connection.  I would like to try to dig in the code but so far am unable to build opensearch security plugin :cry:

**Comment by smortex** [comment]
@dnhodgson I had a similar issue which was patched in #1415 and released in 1.0.1.

Can you test if this fixed your problem and if so close this issue?

**Comment by davidlago** [comment]
We are doing some "spring cleaning in the fall", and to make sure we focus our energies on the right issues and we get a better picture of the state of the repo, we are closing all issues that we are carrying over from the ODFE era (ODFE is no longer supported/maintained, see post [here](https://opendistro.github.io/for-elasticsearch/blog/2022/08/Open-Distro-is-archived/)). 

If you believe this issue should still be considered for current versions of OpenSearch, apologies! Please let us know by re-opening it.

Thanks!
